File: examples/sandbox_client.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.10"
# dependencies = []
# ///
"""
Matchlock Python Client

A simple client for interacting with Matchlock sandbox via JSON-RPC.
Run matchlock in RPC mode: sudo matchlock --rpc

Usage:
    uv run examples/sandbox_client.py
"""

import base64
import json
import os
import subprocess
import sys
from dataclasses import dataclass
from typing import Any


@dataclass
class ExecResult:
    exit_code: int
    stdout: str
    stderr: str
    duration_ms: int


@dataclass
class FileInfo:
    name: str
    size: int
    mode: int
    is_dir: bool


class MatchlockError(Exception):
    """Error from Matchlock RPC"""
    def __init__(self, code: int, message: str):
        self.code = code
        self.message = message
        super().__init__(f"[{code}] {message}")


class MatchlockClient:
    """Client for Matchlock sandbox via JSON-RPC over subprocess stdin/stdout"""
    
    def __init__(self, matchlock_path: str | None = None):
        if matchlock_path is None:
            matchlock_path = os.environ.get("MATCHLOCK_BIN", "./bin/matchlock")
        self.matchlock_path = matchlock_path
        self.process: subprocess.Popen | None = None
        self.request_id = 0
        self.vm_id: str | None = None
    
    def __enter__(self):
        self.start()
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()
    
    def start(self):
        """Start the matchlock RPC process"""
        self.process = subprocess.Popen(
            ["sudo", self.matchlock_path, "--rpc"],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            bufsize=1,
        )
    
    def _send_request(self, method: str, params: dict | None = None) -> Any:
        """Send a JSON-RPC request and return the result"""
        if not self.process or self.process.poll() is not None:
            raise RuntimeError("Matchlock process not running")
        
        self.request_id += 1
        request = {
            "jsonrpc": "2.0",
            "method": method,
            "id": self.request_id,
        }
        if params:
            request["params"] = params
        
        request_json = json.dumps(request)
        self.process.stdin.write(request_json + "\n")
        self.process.stdin.flush()
        
        # Read response (skip event notifications)
        while True:
            line = self.process.stdout.readline()
            if not line:
                raise RuntimeError("Matchlock process closed unexpectedly")
            
            response = json.loads(line)
            
            # Skip notifications (no id field)
            if "id" not in response:
                continue
            
            if response.get("id") != self.request_id:
                continue
            
            if "error" in response and response["error"]:
                err = response["error"]
                raise MatchlockError(err["code"], err["message"])
            
            return response.get("result")
    
    def create(
        self,
        image: str = "standard",
        cpus: int = 1,
        memory_mb: int = 512,
        timeout_seconds: int = 300,
        allowed_hosts: list[str] | None = None,
        mounts: dict[str, dict] | None = None,
    ) -> str:
        """Create and start a new sandbox VM"""
        params = {
            "image": image,
            "resources": {
                "cpus": cpus,
                "memory_mb": memory_mb,
                "timeout_seconds": timeout_seconds,
            },
        }
        
        if allowed_hosts:
            params["network"] = {
                "allowed_hosts": allowed_hosts,
                "block_private_ips": True,
            }
        
        if mounts:
            params["vfs"] = {"mounts": mounts}
        
        result = self._send_request("create", params)
        self.vm_id = result["id"]
        return self.vm_id
    
    def exec(self, command: str, working_dir: str = "") -> ExecResult:
        """Execute a command in the sandbox"""
        params = {"command": command}
        if working_dir:
            params["working_dir"] = working_dir
        
        result = self._send_request("exec", params)
        
        return ExecResult(
            exit_code=result["exit_code"],
            stdout=base64.b64decode(result["stdout"]).decode("utf-8", errors="replace"),
            stderr=base64.b64decode(result["stderr"]).decode("utf-8", errors="replace"),
            duration_ms=result["duration_ms"],
        )
    
    def write_file(self, path: str, content: str | bytes, mode: int = 0o644):
        """Write a file to the sandbox"""
        if isinstance(content, str):
            content = content.encode("utf-8")
        
        params = {
            "path": path,
            "content": base64.b64encode(content).decode("ascii"),
            "mode": mode,
        }
        self._send_request("write_file", params)
    
    def read_file(self, path: str) -> bytes:
        """Read a file from the sandbox"""
        result = self._send_request("read_file", {"path": path})
        return base64.b64decode(result["content"])
    
    def list_files(self, path: str) -> list[FileInfo]:
        """List files in a directory"""
        result = self._send_request("list_files", {"path": path})
        return [
            FileInfo(
                name=f["name"],
                size=f["size"],
                mode=f["mode"],
                is_dir=f["is_dir"],
            )
            for f in result.get("files", [])
        ]
    
    def close(self):
        """Close the sandbox and cleanup"""
        if self.process and self.process.poll() is None:
            try:
                self._send_request("close")
            except Exception:
                pass
            try:
                self.process.terminate()
                self.process.wait(timeout=2)
            except Exception:
                self.process.kill()
                self.process.wait(timeout=1)
        self.vm_id = None


def main():
    """Example usage of the Matchlock client"""
    print("=== Matchlock Python Client Example ===\n")
    
    with MatchlockClient() as client:
        # Create a sandbox
        print("Creating sandbox...")
        vm_id = client.create(
            image="standard",
            cpus=1,
            memory_mb=512,
        )
        print(f"Created VM: {vm_id}\n")
        
        # Execute a simple command
        print("Running 'echo Hello from sandbox!'...")
        result = client.exec("echo 'Hello from sandbox!'")
        print(f"  stdout: {result.stdout.strip()}")
        print(f"  exit_code: {result.exit_code}")
        print(f"  duration: {result.duration_ms}ms\n")
        
        # Write a Python script to the sandbox
        script = '''
import sys
print(f"Python version: {sys.version}")
print("Arguments:", sys.argv[1:])
for i in range(3):
    print(f"Count: {i}")
'''
        print("Writing Python script to /workspace/test.py...")
        client.write_file("/workspace/test.py", script)
        
        # Execute the script
        print("Running the script...")
        result = client.exec("python3 /workspace/test.py arg1 arg2")
        print(f"  stdout:\n{result.stdout}")
        if result.stderr:
            print(f"  stderr: {result.stderr}")
        print(f"  exit_code: {result.exit_code}\n")
        
        # List files
        print("Listing /workspace...")
        files = client.list_files("/workspace")
        for f in files:
            ftype = "dir" if f.is_dir else "file"
            print(f"  {f.name} ({ftype}, {f.size} bytes)")
        print()
        
        # Read the file back
        print("Reading /workspace/test.py back...")
        content = client.read_file("/workspace/test.py")
        print(f"  Content length: {len(content)} bytes")
        print(f"  First line: {content.decode().split(chr(10))[0]}\n")
        
        # Test error handling
        print("Testing command that fails...")
        result = client.exec("exit 42")
        print(f"  exit_code: {result.exit_code}\n")
        
        print("Closing sandbox...")
    
    print("Done!")


if __name__ == "__main__":
    main()

File: examples/sandbox_client.go
// Matchlock Go Client Example
//
// A simple client for interacting with Matchlock sandbox via JSON-RPC.
//
// Usage:
//
//	go run examples/sandbox_client.go
package main

import (
	"bufio"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"sync/atomic"
)

// Request represents a JSON-RPC request
type Request struct {
	JSONRPC string      `json:"jsonrpc"`
	Method  string      `json:"method"`
	Params  interface{} `json:"params,omitempty"`
	ID      uint64      `json:"id"`
}

// Response represents a JSON-RPC response
type Response struct {
	JSONRPC string          `json:"jsonrpc"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   *RPCError       `json:"error,omitempty"`
	ID      *uint64         `json:"id,omitempty"`
}

// RPCError represents a JSON-RPC error
type RPCError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

func (e *RPCError) Error() string {
	return fmt.Sprintf("[%d] %s", e.Code, e.Message)
}

// ExecResult represents the result of command execution
type ExecResult struct {
	ExitCode   int
	Stdout     string
	Stderr     string
	DurationMS int64
}

// FileInfo represents file metadata
type FileInfo struct {
	Name  string `json:"name"`
	Size  int64  `json:"size"`
	Mode  uint32 `json:"mode"`
	IsDir bool   `json:"is_dir"`
}

// Client is a Matchlock JSON-RPC client
type Client struct {
	cmd       *exec.Cmd
	stdin     io.WriteCloser
	stdout    *bufio.Reader
	requestID atomic.Uint64
	vmID      string
}

// NewClient creates a new Matchlock client
func NewClient(matchlockPath string) (*Client, error) {
	cmd := exec.Command("sudo", matchlockPath, "--rpc")

	stdin, err := cmd.StdinPipe()
	if err != nil {
		return nil, fmt.Errorf("failed to get stdin pipe: %w", err)
	}

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, fmt.Errorf("failed to get stdout pipe: %w", err)
	}

	cmd.Stderr = os.Stderr

	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("failed to start matchlock: %w", err)
	}

	return &Client{
		cmd:    cmd,
		stdin:  stdin,
		stdout: bufio.NewReader(stdout),
	}, nil
}

// sendRequest sends a JSON-RPC request and returns the result
func (c *Client) sendRequest(method string, params interface{}) (json.RawMessage, error) {
	id := c.requestID.Add(1)

	req := Request{
		JSONRPC: "2.0",
		Method:  method,
		Params:  params,
		ID:      id,
	}

	data, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	if _, err := fmt.Fprintln(c.stdin, string(data)); err != nil {
		return nil, fmt.Errorf("failed to write request: %w", err)
	}

	// Read response (skip notifications)
	for {
		line, err := c.stdout.ReadBytes('\n')
		if err != nil {
			return nil, fmt.Errorf("failed to read response: %w", err)
		}

		var resp Response
		if err := json.Unmarshal(line, &resp); err != nil {
			return nil, fmt.Errorf("failed to unmarshal response: %w", err)
		}

		// Skip notifications (no ID)
		if resp.ID == nil {
			continue
		}

		if *resp.ID != id {
			continue
		}

		if resp.Error != nil {
			return nil, resp.Error
		}

		return resp.Result, nil
	}
}

// CreateConfig holds configuration for creating a sandbox
type CreateConfig struct {
	Image          string
	CPUs           int
	MemoryMB       int
	TimeoutSeconds int
	AllowedHosts   []string
}

// Create creates and starts a new sandbox VM
func (c *Client) Create(cfg CreateConfig) (string, error) {
	if cfg.Image == "" {
		cfg.Image = "standard"
	}
	if cfg.CPUs == 0 {
		cfg.CPUs = 1
	}
	if cfg.MemoryMB == 0 {
		cfg.MemoryMB = 512
	}
	if cfg.TimeoutSeconds == 0 {
		cfg.TimeoutSeconds = 300
	}

	params := map[string]interface{}{
		"image": cfg.Image,
		"resources": map[string]interface{}{
			"cpus":            cfg.CPUs,
			"memory_mb":       cfg.MemoryMB,
			"timeout_seconds": cfg.TimeoutSeconds,
		},
	}

	if len(cfg.AllowedHosts) > 0 {
		params["network"] = map[string]interface{}{
			"allowed_hosts":    cfg.AllowedHosts,
			"block_private_ips": true,
		}
	}

	result, err := c.sendRequest("create", params)
	if err != nil {
		return "", err
	}

	var createResult struct {
		ID string `json:"id"`
	}
	if err := json.Unmarshal(result, &createResult); err != nil {
		return "", fmt.Errorf("failed to parse create result: %w", err)
	}

	c.vmID = createResult.ID
	return c.vmID, nil
}

// Exec executes a command in the sandbox
func (c *Client) Exec(command string) (*ExecResult, error) {
	params := map[string]string{
		"command": command,
	}

	result, err := c.sendRequest("exec", params)
	if err != nil {
		return nil, err
	}

	var execResult struct {
		ExitCode   int    `json:"exit_code"`
		Stdout     string `json:"stdout"`
		Stderr     string `json:"stderr"`
		DurationMS int64  `json:"duration_ms"`
	}
	if err := json.Unmarshal(result, &execResult); err != nil {
		return nil, fmt.Errorf("failed to parse exec result: %w", err)
	}

	stdout, _ := base64.StdEncoding.DecodeString(execResult.Stdout)
	stderr, _ := base64.StdEncoding.DecodeString(execResult.Stderr)

	return &ExecResult{
		ExitCode:   execResult.ExitCode,
		Stdout:     string(stdout),
		Stderr:     string(stderr),
		DurationMS: execResult.DurationMS,
	}, nil
}

// WriteFile writes a file to the sandbox
func (c *Client) WriteFile(path string, content []byte, mode uint32) error {
	if mode == 0 {
		mode = 0644
	}

	params := map[string]interface{}{
		"path":    path,
		"content": base64.StdEncoding.EncodeToString(content),
		"mode":    mode,
	}

	_, err := c.sendRequest("write_file", params)
	return err
}

// ReadFile reads a file from the sandbox
func (c *Client) ReadFile(path string) ([]byte, error) {
	params := map[string]string{
		"path": path,
	}

	result, err := c.sendRequest("read_file", params)
	if err != nil {
		return nil, err
	}

	var readResult struct {
		Content string `json:"content"`
	}
	if err := json.Unmarshal(result, &readResult); err != nil {
		return nil, fmt.Errorf("failed to parse read result: %w", err)
	}

	return base64.StdEncoding.DecodeString(readResult.Content)
}

// ListFiles lists files in a directory
func (c *Client) ListFiles(path string) ([]FileInfo, error) {
	params := map[string]string{
		"path": path,
	}

	result, err := c.sendRequest("list_files", params)
	if err != nil {
		return nil, err
	}

	var listResult struct {
		Files []FileInfo `json:"files"`
	}
	if err := json.Unmarshal(result, &listResult); err != nil {
		return nil, fmt.Errorf("failed to parse list result: %w", err)
	}

	return listResult.Files, nil
}

// Close closes the sandbox and cleans up
func (c *Client) Close() error {
	c.sendRequest("close", nil)
	c.stdin.Close()
	return c.cmd.Wait()
}

func main() {
	fmt.Println("=== Matchlock Go Client Example ===\n")

	// Create client - use ./bin/matchlock or MATCHLOCK_BIN env var
	matchlockBin := os.Getenv("MATCHLOCK_BIN")
	if matchlockBin == "" {
		matchlockBin = "./bin/matchlock"
	}
	client, err := NewClient(matchlockBin)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create client: %v\n", err)
		os.Exit(1)
	}
	defer client.Close()

	// Create sandbox
	fmt.Println("Creating sandbox...")
	vmID, err := client.Create(CreateConfig{
		Image:    "standard",
		CPUs:     1,
		MemoryMB: 512,
	})
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create sandbox: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Created VM: %s\n\n", vmID)

	// Execute a simple command
	fmt.Println("Running 'echo Hello from sandbox!'...")
	result, err := client.Exec("echo 'Hello from sandbox!'")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to exec: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("  stdout: %s", result.Stdout)
	fmt.Printf("  exit_code: %d\n", result.ExitCode)
	fmt.Printf("  duration: %dms\n\n", result.DurationMS)

	// Write a script to the sandbox
	script := `#!/bin/sh
echo "Hello from script!"
echo "Current directory: $(pwd)"
echo "Files in /workspace:"
ls -la /workspace
`
	fmt.Println("Writing script to /workspace/test.sh...")
	if err := client.WriteFile("/workspace/test.sh", []byte(script), 0755); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write file: %v\n", err)
		os.Exit(1)
	}

	// Execute the script
	fmt.Println("Running the script...")
	result, err = client.Exec("sh /workspace/test.sh")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to exec script: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("  stdout:\n%s", result.Stdout)
	if result.Stderr != "" {
		fmt.Printf("  stderr: %s\n", result.Stderr)
	}
	fmt.Printf("  exit_code: %d\n\n", result.ExitCode)

	// List files
	fmt.Println("Listing /workspace...")
	files, err := client.ListFiles("/workspace")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to list files: %v\n", err)
		os.Exit(1)
	}
	for _, f := range files {
		ftype := "file"
		if f.IsDir {
			ftype = "dir"
		}
		fmt.Printf("  %s (%s, %d bytes)\n", f.Name, ftype, f.Size)
	}
	fmt.Println()

	// Read the file back
	fmt.Println("Reading /workspace/test.sh back...")
	content, err := client.ReadFile("/workspace/test.sh")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to read file: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("  Content length: %d bytes\n", len(content))
	fmt.Printf("  First line: %s\n\n", string(content[:20]))

	// Test error handling
	fmt.Println("Testing command that fails...")
	result, err = client.Exec("exit 42")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Exec returned error: %v\n", err)
	} else {
		fmt.Printf("  exit_code: %d\n\n", result.ExitCode)
	}

	fmt.Println("Closing sandbox...")
	fmt.Println("Done!")
}
